src/metldata/builtin_transformations/count_content_values/data_transform.py

```diff
@@ -0,0 +1,132 @@
+# limitations under the License.
+
+"""Data transformation logic for count content values transformation."""
+
+from collections import Counter
+
+from schemapack.spec.datapack import DataPack
+
+from metldata.builtin_transformations.add_content_properties.path import (
+    resolve_data_object_path,
+)
+from metldata.builtin_transformations.common.path.path_utils import (
+    get_directly_referenced_class,
+)
+from metldata.builtin_transformations.count_content_values.instruction import (
+    CountContentValueInstruction,
+)
+from metldata.transform.exceptions import (
+    EvitableTransformationError,
+)
+
+
+class TransformationContext:
+    """Context manager for a data transformation process, wrapping the Datapack to be
+    transformed and transformation instructions per class.
+    """
+
+    def __init__(
+        self,
+        data: DataPack,
+        instructions_by_class: dict[str, list[CountContentValueInstruction]],
+    ):
+        self.data = data.model_copy(deep=True)
+        self.instructions_by_class = instructions_by_class
+
+    def get_class_resources(self, class_name: str):
+        """Extract the resources of a given class from a Datapack."""
+        resources = self.data.resources.get(class_name)
+        if not resources:
+            raise EvitableTransformationError()
+        return resources
+
+
+def count_content(
+    *,
+    data: DataPack,
+    instructions_by_class: dict[str, list[CountContentValueInstruction]],
+) -> DataPack:
+    """Transforms the data."""
+    context = TransformationContext(data, instructions_by_class)
+    for class_name, instructions in context.instructions_by_class.items():
+        resources = context.get_class_resources(class_name)
+
+        for instruction in instructions:
+            relation_path = instruction.source.relation_path
+            referenced_class = get_directly_referenced_class(relation_path)
+
+            # Only one element is expected in the path
+            relation_name = relation_path.elements[0].property
+            # content_resources keeping the values to be count are obtained from
+            # the class that is referenced by the transformation target class through
+            # relation_name.
+            content_resources = context.get_class_resources(referenced_class)
+
+            transform_resources(
+                resources, relation_name, content_resources, instruction
+            )
+
+    return context.data
+
+
+def transform_resources(
+    resources,
+    relation_name,
+    content_resources,
+    instruction: CountContentValueInstruction,
+):
+    """Transform resources"""
+    for resource in resources.values():
+        content = resource.content
+        related_to = resource.relations.get(relation_name)
+        count_values = get_count_values(related_to, content_resources, instruction)
+
+        target_object = get_target_object(content, instruction)
+        target_object[instruction.target_content.property_name] = dict(
```

Comment:
Would factor `instruction.target_content.property_name` out into its own variable for readability purposes, even if we don't reuse it